lib/clib/CCorresLemmas.thy

```diff
@@ -586,6 +586,13 @@ lemma ccorres_assert2:
               (\<lambda>s. P \<longrightarrow> G s) {s. P \<longrightarrow> s \<in> G'} hs (assert P >>= f) c"
   by (cases P, simp_all add: ccorres_fail')
 
+(* A weaker version of ccorres_assert2 that may be easier to match with schematics, since
+   the concrete guard does not mention the asserted property *)
+lemma ccorres_assert2':
```

Comment:
Maybe this could have a more informative name. I could rename regular `ccorres_assert` to `ccorres_assert_drop_property` or something like that, `ccorres_assert2` to `ccorres_assert_add_property` or something, and then this new rule could maybe be `ccorres_assert_add_abstract_property`. I'm not married to these names, these are just suggestions to indicate what's happening

Reply:
Hm, this is a bit subtle. The default `corres_assert` doesn't actually drop the property, it does provide it in the subgoal, and it is a bit strange that we do sometimes seem to need it in the generated precondition -- ideally the sub proof should not produce anything that needs `P` any more, because `P` was available already and should have been eliminated.

That said, the way the schematics are instantiated, we often do not want to `simp` or similar in the sub proof, so I agree that we do want these variants. The problem (as you seem to have discovered if you have an abstract-only variant) is that adding implications to the preconditions can make subsequent proofs harder even though they are logically weaker/easier.

Of course the current variant names are terrible. I would propose the following scheme: `ccorres_assert` should stay the default name, and we add `ccorres_assert_both` + `ccorres_assert_abs`. If we ever need it, we could add `ccorres_assert_C` (or `_concrete` or similar).

Reply:
This is a bit tricky to figure out. After I wrote the original comment about add and dropping properties I wasn't so sure why it is that it seems like the property that's asserted disappears sometimes. Are you thinking that it's because it's `simp`ed away? And that having it in the guard explicitly preserves it until the final goal where the guard implications are resolved?

The reason I wanted this variant was that `ccorres_symb_exec_l'` won't change the schematic, and so if we use `ccorres_symb_exec_l'` and then want to use an assert rule about the value that we get, then we can't unify with `ccorres_assert2`.

Reply:
This is indeed strange, I feel like we shouldn't need this rule. We generally rewrite stuff on the C side into Haskell side as we symbolically execute forward, meaning that an assertion on the Haskell side might later provide information to clean up the final C part of the proof, so it should be available to both sides, and if it isn't relevant, it shouldn't get in the way.

It's also possible that the `ccorres_symb_exec_l'` isn't in the right form. I'm guessing none of the other `symb_exec_l` rules worked for you?

The property "disappearing" sounds like it's getting simped away, but it persists in the guards.

If you want to keep this rule, I'd suggest changing the name to `ccorres_assert2_abs` and changing the comment to something like `(* This variant only propagates the assertion into the abstract side, to aid with schematic matching on the concrete side. *)`

Reply:
> The reason I wanted this variant was that ccorres_symb_exec_l' won't change the schematic, and so if we use ccorres_symb_exec_l' and then want to use an assert rule about the value that we get, then we can't unify with ccorres_assert2.

Right, that makes sense to me. Together with Raf's proposed comment that should resolve the content at least.

> After I wrote the original comment about add and dropping properties I wasn't so sure why it is that it seems like the property that's asserted disappears sometimes. Are you thinking that it's because it's simped away? 

It's scope: the default rule makes `P` available in the assumption of all sub goals that are "inside" the scope of the assert rule applications (all new subgoals). The assert2 rule exports `P` additionally to the C vcg goal that is on the same level as the assert2 rule application -- the vcg goal already exists when the rule is being applied and the goal is being modified by partially instantiating the schematic. This in turn usually make `P` available in in the big final goal.

Usually that second part shouldn't be necessary, because you should need the `P` only for code that comes after the `assert`, which means in the subgoals of the assert rule. But sometimes that fails, because we do export `P`-dependent conditions in those vcg goals and sometimes also abstract goals.